### ISSUE TITLE
Mouse-controlled camera rotation

### ISSUE THREAD

**Issue by tomluchowski** [OpenDungeons/OpenDungeons#412]
Hello , do you wish to have some of Autoscroll and autorotate ; or the keyboard scrolling is going to stay forever ? Does the 

> This was coming from the fact of using the OS cursor only, as CEGUI considers
> it being 0x0 in size, and then computes the offset badly in that case.
> I thus added an empty 12x12 cursor and used it without hiding the system one."

would help to have correct margins set ?

**Comment by hwoarangmy** [comment]
Well, I guess a user friendly autoscroll is always welcome. Concerning your quote, it is about using a margin for tooltips. But it may have something to do with what you have in mind.

**Comment by akien-mga** [comment]
I can visualise the autoscroll (you move the mouse to the sides of the screen and it moves the map), but how would autorotate work?

I'm favourable to implementing the autoscroll though, maybe with a setting to enable/disable it in the options menu (probably enabled by default).

**Comment by Bertram25** [comment]
Just as a warning notice, the last time, @tomluchowski removed the keyboard movement and added a (unfortunately somewhat buggy) autoscroll feature, which is why I removed it in the first place.

So as a first point, adding the auto scroll or not, the keyboard movement shall stay, I, for one, couldn't play without it.

Secondly, it is all a question of taste, but auto-scroll when moving the cursor on the screen edges was rather annoying when you were selecting stuff around, at least the way it was done.
Ok, as a an option but I'd request it to be a slider to set up its sensitivity in the options.

The issue would then have the prerequisite of #3

**Comment by tomluchowski** [comment]
AFAIR keyboard movement was left. 
Tough there are some nasty bugs when setting the 

"paramList.insert(std::make_pair(std::string("x11_mouse_grab"), std::string("true")));"

This seems to be necessary to have force the mouse pointer to reach 0 when on max to the left and window.width when max to the right ... 
EDIT : sorry my issues were : the window didn;'t got autofocus due to my window manager settings ...
Could we make it to always steal the focus when running in fullscreen mode ?

**Comment by tomluchowski** [comment]
The camera movement is realized by quite complicated function , then you are saying mine scroll is too fast or to awkward , indeed calibrating it would require time, does anyone understands what's going in the code below ? Maybe we should rewrite that function from scratch .... 
BTW: 
How does one paste the code in here ? 
http://pastebin.com/NhTT55GT

**Comment by tomluchowski** [comment]
I have restored what I've got to restore, people who hasn't tried it yet please try it.
Now to rotate the camera you can also : click the LMB on the screen  and drag cursor ( with LMB pressed ) along the nearest screen edge ... 
 After reading the code I now got what Bertram meant that it interfers with creature Drag and drop .... maybe someone have suggestions ...

**Comment by Bertram25** [comment]
Hey @tomluchowski :) And Happy New Year, btw.

> EDIT : sorry my issues were : the window didn;'t got autofocus due to my window manager settings ...
> Could we make it to always steal the focus when running in fullscreen mode ?

I really don't know. I guess the actual need would then be to dynamically grab mouse input when in fullscreen mode in X11, right?

Anyway, changing OD code to counter your window manager settings isn't the way to go as you explicitely told your system not to give the focus to an app. So no wonder it does exactly what's told, right?

Also, we've added the following option:
https://github.com/OpenDungeons/OpenDungeons/blob/development/CMakeLists.txt#L58
`option(OD_LINUX_NO_KEYBOARD_GRAB "Do not grab keyboard input when using OIS under Linux. Permits input in other applications." ON)`
If you set this option to false when invoking cmake, such as:
`cmake -DOD_LINUX_NO_KEYBOARD_GRAB=off .` for instance, OD will grab the input again when starting.

I don't think it's a good idea to do more in the project code about this, at least for the reasons you gave.

> I have restored what I've got to restore, people who hasn't tried it yet please try it.
> Now to rotate the camera you can also : click the LMB on the screen and drag cursor ( with LMB pressed ) along the nearest screen edge ...
> After reading the code I now got what Bertram meant that it interfers with creature Drag and drop .... maybe someone have suggestions ... 

Note that I've simplified the camera update code in https://github.com/OpenDungeons/OpenDungeons/commit/9943badae3ccdb1ae3d4481000e666dd4e4d2d55 just in case.

To be able to tell you more, I guess we'll need to see what you've made in a PR, so we'll be able to test it, and comment accordingly.

**Comment by akien-mga** [comment]
> To be able to tell you more, I guess we'll need to see what you've made, so we'll be able to test it, and comment accordingly.

@tomluchowski has pushed his work on his own fork: https://github.com/tomluchowski/OpenDungeons

I just tested it, basically:
- Mouse scrolling works fine, the added functionality doesn't seem to impact the keyboard scrolling.
- The scrolling is too fast, it's hard to adjust over your base with it IMO. I would reduce the speed at least by a factor 2.
- I'm not convinced about the camera rotation with the mouse, as pointed out it makes an important part of the screen unusable gameplay-wise (you can't click units or mark tiles, because the left and right margins are used to rotate the camera). I think camera rotation should be made with another kind of input not used in game, maybe right click + move the cursor?
- The forced mouse grab is a bother to me. I'm used to playing in windowed mode so that I can commit on Github at the same time, and with the mouse grab I can't get the mouse back in my WM.

About the mouse grab, I think we should only implement @tomluchowski's code once we have a settings manager, and let it "off" by default, so that people starting the game in windowed mode don't get confused. For me the "optimal" setting would be:
- In windowed mode, no mouse grab (so no autoscroll).
- In fullscreen mode, mouse grab + autoscroll.

**Comment by Bertram25** [comment]
As for me, I can't get why using simple keys Q and E on a US keyboard as it is now doesn't suffice to rotate the screen.
It is as if @tomluchowski was playing OD with only one hand and it doesn't make sense to me.

**Comment by tomluchowski** [comment]
Actually one of the Peter's Molenoux games -- Black and White set and achieved very ambitious goal of having the game bbeing controlled only by mouse ( without the icons BTW !!!! ) and after some litllte learing it turned out to be very handful and nice for me for that example ....
I can type quite fast , though I am not notorious gamer I am not just used to play both hands ( and being an shorcut keys virtuoze -- as starcraft lunatics giving more than 350 orders per minute ) .
Giving the player a chance  playing the game through and forgetting the keyboard --- that might be a great goal :)

@akien-mga I don't care about mouse-grab anymore, all that matters is that the window have the focus from the OS and I seemed to be quite confused about that ....

**Comment by hwoarangmy** [comment]
> It is as if @tomluchowski was playing OD with only one hand and it doesn't make sense to me

Or he may be preparing an Android version ^^
Anyway, if it works, it may be interesting for people who want to play with mouse only. We just have to disable it until we have the option menu.

**Comment by Bertram25** [comment]
@tomluchowski 
Ok, this is making more sense now. This is definitely not something standard but I understand better the will to add this. The need of an option menu is ever growing IMHO. I definitely need to tackle that one down. ;)

@hwoarangmy 

> Or he may be preparing an Android version ^^

Eh eh. This is a fair point, but I do think the Android version will need another type of input. That said, there is no reason not to add it, especially if that can be an option. :)

**Comment by tomluchowski** [comment]
I could implement a set of console's cmds to set whatever you wish , problem is whether you would mericifully merge the result of my work into main trunk

**Comment by Bertram25** [comment]
> I could implement a set of console's cmds to set whatever you wish , problem is whether you would mericifully merge the result of my work into main trunk 

It depends on you, not on me, @tomluchowski . No, don't play the victim with me.
It's simple:
- Discuss what you want to add first and see whether everyone is agree before starting crazy things.
- Follow the code conventions.
- Do a Pull request and try to fix the code according to everyone comments.

Then, once the code is working and looking sane, it will be merged by one of the reviewers.

Also, the issue here is not to add a set of commands AFAIK but to add a sane way to autoscroll and autorotate, right?

**Comment by akien-mga** [comment]
Note that the "autorotate" part of this issue has been implementend in #892. The changes regarding the scrolling have been reverted for now in this same PR.

For clarity, I'll propose to close this issue as fixed once #892 is merged, and open a new issue to discuss the scrolling behaviour that we would want to see implemented.

**Comment by hwoarangmy** [comment]
as @akien-mga said, this one is closed
